kubectl-plugin/pkg/cmd/cluster/log.go

```diff
@@ -0,0 +1,161 @@
+	cmdFactory := cmdutil.NewFactory(options.configFlags)
+
+	cmd := &cobra.Command{
+		Use:          "log (CLUSTER_NAME)",
+		Short:        "Get cluster log",
+		Aliases:      []string{"logs"},
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := options.Complete(args); err != nil {
+				return err
+			}
+			if err := options.Validate(); err != nil {
+				return err
+			}
+			return options.Run(cmd.Context(), cmdFactory)
+		},
+	}
+	cmd.Flags().StringVar(&options.outputDir, "out-dir", options.outputDir, "File Directory PATH of where to download the file logs to.")
+	options.configFlags.AddFlags(cmd.Flags())
+	return cmd
+}
+
+func (options *ClusterLogOptions) Complete(args []string) error {
+	if options.outputDir != "" {
+		options.downloadLog = true
+	}
+
+	options.args = args
+	return nil
+}
+
+func (options *ClusterLogOptions) Validate() error {
+	// Overrides and binds the kube config then retrieves the merged result
+	config, err := options.configFlags.ToRawKubeConfigLoader().RawConfig()
+	if err != nil {
+		return fmt.Errorf("Error retrieving raw config: %w", err)
+	}
+	if len(config.CurrentContext) == 0 {
+		return fmt.Errorf("no context is currently set, use %q to select a new one", "kubectl config use-context <context>")
+	}
+	// If there is a output directory to download to, we validate the directory
+	if options.downloadLog {
+		info, err := os.Stat(options.outputDir)
+		if os.IsNotExist(err) {
+			return fmt.Errorf("Directory does not exist. Failed with %w", err)
+		} else if err != nil {
+			return fmt.Errorf("Error occurred will checking directory %w", err)
+		} else if !info.IsDir() {
+			return fmt.Errorf("Path is Not a directory. Please input a directory and try again")
+		}
+	}
+	// Command must have ray cluster name
+	if len(options.args) > 1 {
+		return fmt.Errorf("must have at most one argument")
+	}
+	return nil
+}
+
+func (options *ClusterLogOptions) Run(ctx context.Context, factory cmdutil.Factory) error {
+	kubeClientSet, err := factory.KubernetesClientSet()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve kubernetes client set: %w", err)
+	}
+
+	var listopts v1.ListOptions
+	if len(options.args) == 1 {
+		listopts = v1.ListOptions{
+			LabelSelector: fmt.Sprintf("ray.io/group=headgroup, ray.io/cluster=%s", options.args[0]),
+		}
+	} else {
+		listopts = v1.ListOptions{
+			LabelSelector: "ray.io/group=headgroup",
+		}
+	}
+
+	// Get list of nodes that are considered ray heads
+	rayHeads, err := kubeClientSet.CoreV1().Pods(*options.configFlags.Namespace).List(ctx, listopts)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve head node for cluster %s: %w", options.args[0], err)
+	}
+
+	// get a list of logs of the ray heads.
+	var logList []*bytes.Buffer
+	for _, rayHead := range rayHeads.Items {
+		request := kubeClientSet.CoreV1().Pods(rayHead.Namespace).GetLogs(rayHead.Name, &corev1.PodLogOptions{})
```

Comment:
I don't think it's enough to just download the stdout of the Ray head. We'll also want to save logs in `/tmp/ray/session_latest/logs` too. What do you think @kevin85421 @MortalHappiness?

Reply:
Hmmm that makes sense. If thats the case, then the `out-dir` should be a requirement for the command then? Or do we still want the ability to print out all the logs under `/tmp/ray/session_latest/logs`

Reply:
out-dir being required makes sense to me. I don't think we need to print out all the logs, saving to out-dir as the default makes sense!

Reply:
I also think both the log of the Ray head and the Ray workers should be downloaded. If displaying to stdout is needed, we can print some additional information like what [bat](https://github.com/sharkdp/bat) does:

```
$ kubectl ray cluster logs my-cluster

------------
raycluster-sample-head-wnsmd
------------
(the log of ray head)

------------
raycluster-sample-worker-small-group-9j7h7
------------
(the log of ray worker)
```

```
$ kubectl ray cluster logs my-cluster --out-dir ./log-dir
$ ls ./log-dir

raycluster-sample-head-wnsmd.log raycluster-sample-worker-small-group-9j7h7.log
```

@andrewsykim Is what I mentioned above the result you originally intended to implement?

Reply:
Make sense, but maybe each head / worker pod should be a directory, and inside each directory is all the files in `/tmp/ray/session_latest/logs`

Reply:
I'm okay scoping this PR to only the head pod since it should be easy to extend it later to also donwload logs for the worker pods

Reply:
> I don't think it's enough to just download the stdout of the Ray head. We'll also want to save logs in /tmp/ray/session_latest/logs too.

+1, we propose this function because Ray is not very K8s-friendly. Multiple processes are running in a single container and writing logs to files under /tmp/ray/session_latest instead of STDOUT/STDERR.